File: Sources/RichText/Find/RichTextFindBar.swift
// Sources/RichText/Find/RichTextFindBar.swift
//
// The standalone find UI (layer 3, see RichTextSearch.swift): a bar that edits a RichTextFindController,
// and two modifiers that wire it up - `RichText.find(_:)` binds a document to the controller and paints its
// matches, `View.richTextFindBar(_:)` shows the bar on a container and installs the keyboard shortcuts.
//
//   @StateObject private var find = RichTextFindController()
//   ScrollView { RichText(document).find(find) }
//       .richTextFindBar(find)
//
// A host that has its own search field skips both the bar and the shortcuts and just sets
// `controller.query`; `RichText.find(_:)` still paints. The bar is deliberately plain - a text field, the
// count, previous / next, an options menu and a close button - so it can sit inside any chrome.

import SwiftUI

public struct RichTextFindBar: View {
    @ObservedObject private var controller: RichTextFindController
    @FocusState private var fieldFocused: Bool
    private let placeholder: String

    public init(_ controller: RichTextFindController, placeholder: String = "Find") {
        self.controller = controller
        self.placeholder = placeholder
    }

    public var body: some View {
        HStack(spacing: 6) {
            Image(systemName: "magnifyingglass")
                .foregroundStyle(.secondary)
            TextField(placeholder, text: $controller.query)
                .textFieldStyle(.plain)
                .focused($fieldFocused)
                .onSubmit {
                    controller.next()
                }
                #if os(iOS)
                .autocorrectionDisabled()
                .textInputAutocapitalization(.never)
                #endif
            Text(controller.summary)
                .font(.callout.monospacedDigit())
                .foregroundStyle(.secondary)
                .lineLimit(1)
                .fixedSize()
            Button {
                controller.previous()
            } label: {
                Image(systemName: "chevron.up")
            }
            .keyboardShortcut("g", modifiers: [.command, .shift])
            .disabled(controller.ranges.isEmpty)
            .accessibilityLabel("Previous match")
            Button {
                controller.next()
            } label: {
                Image(systemName: "chevron.down")
            }
            .keyboardShortcut("g", modifiers: .command)
            .disabled(controller.ranges.isEmpty)
            .accessibilityLabel("Next match")
            Menu {
                Toggle("Match Case", isOn: $controller.options.caseSensitive)
                Toggle("Whole Words", isOn: $controller.options.wholeWord)
                Toggle("Match Diacritics", isOn: $controller.options.diacriticSensitive)
                Toggle("Regular Expression", isOn: $controller.options.regularExpression)
            } label: {
                Image(systemName: "ellipsis.circle")
            }
            .menuIndicator(.hidden)
            .fixedSize()
            .accessibilityLabel("Find options")
            Button {
                controller.dismiss()
            } label: {
                Image(systemName: "xmark.circle.fill")
            }
            .keyboardShortcut(.escape, modifiers: [])
            .accessibilityLabel("Close find")
        }
        .buttonStyle(.borderless)
        .padding(.horizontal, 10)
        .padding(.vertical, 6)
        .background(.bar)
        // Focus when a Cmd-F asked for it and no bar has honored that request yet: a repeat Cmd-F while
        // the bar is open puts the reader back in the field, a bar presented by a host (`present(focus:
        // false)`) leaves focus in the host's own field.
        .onAppear {
            focusIfRequested()
        }
        .onChange(of: controller.focusRequests) { _ in
            focusIfRequested()
        }
    }

    private func focusIfRequested() {
        guard controller.focusRequests != controller.focusHonored else {
            return
        }
        controller.markFocusHonored()
        fieldFocused = true
    }
}

public extension RichText {
    /// Paint `controller`'s matches on this document and keep the controller bound to this document's
    /// rendered text. The view that owns the controller (`@StateObject` / `@ObservedObject`) re-renders on
    /// every controller change, which is what re-evaluates this call with fresh highlights.
    ///
    /// Returns `some View`, so apply RichText's own modifiers (`widthBehavior`) BEFORE it. Because that
    /// owning view re-renders per keystroke, build the document's attributed string once and pass it
    /// through `RichText(attributed:)` (as the demo does) rather than re-rendering Markdown in `body`.
    /// When the rendered text changes, the previous ranges may paint for one frame until the controller
    /// re-binds (out-of-range ones are clipped, never fatal).
    func find(_ controller: RichTextFindController) -> some View {
        RichTextFindBinding(controller: controller, text: renderedText, content: findHighlights(controller.highlights))
    }
}

// Binds on appear and whenever the rendered text changes (a re-rendered document), never during a body
// evaluation, so the controller's published state is not mutated mid-update.
private struct RichTextFindBinding<Content: View>: View {
    let controller: RichTextFindController
    let text: String
    let content: Content

    var body: some View {
        content
            .onAppear {
                controller.bind(text: text)
            }
            .onChange(of: text) { newText in
                controller.bind(text: newText)
            }
    }
}

public extension View {
    /// Show `RichTextFindBar` on this container while `controller.isPresented`, inset at `edge`, and
    /// install Cmd-F to present it. Pair with `RichText.find(_:)` on the document inside.
    func richTextFindBar(_ controller: RichTextFindController, edge: VerticalEdge = .top,
                         placeholder: String = "Find") -> some View {
        modifier(RichTextFindBarModifier(controller: controller, edge: edge, placeholder: placeholder))
    }
}

private struct RichTextFindBarModifier: ViewModifier {
    @ObservedObject var controller: RichTextFindController
    let edge: VerticalEdge
    let placeholder: String

    func body(content: Content) -> some View {
        content
            .safeAreaInset(edge: edge, spacing: 0) {
                if controller.isPresented {
                    RichTextFindBar(controller, placeholder: placeholder)
                }
            }
            .background {
                // An invisible button carries the shortcut; a Button is the one SwiftUI element whose
                // keyboardShortcut works on every platform without a menu.
                Button("Find") {
                    controller.present()
                }
                .keyboardShortcut("f", modifiers: .command)
                .hidden()
            }
    }
}

File: android/richtext/src/main/kotlin/com/abracode/richtext/find/RichTextFindBar.kt
package com.abracode.richtext.find

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.Checkbox
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp

// Port of Sources/RichText/Find/RichTextFindBar.swift - the standalone find UI (layer 3): a bar that edits a
// RichTextFindController. The host places it (a Column above the document, a top app bar slot) and shows it while
// `controller.isPresented`; on Android the reader opens it from a host affordance (an app-bar search icon that
// calls `controller.present()`), there being no Cmd-F. A host with its own search field skips the bar and sets
// `controller.query`; `RichText(find = controller)` still paints.

/** The find bar: field, "n of m", previous / next, an options menu and close. */
@Composable
fun RichTextFindBar(controller: RichTextFindController, modifier: Modifier = Modifier, placeholder: String = "Find") {
    val focusRequester = remember { FocusRequester() }
    var menuOpen by remember { mutableStateOf(false) }
    // Focus when a present(focus = true) asked for it and no bar has honored that request yet: a repeat request
    // while the bar is open puts the reader back in the field, a host-driven present leaves focus in the host's field.
    LaunchedEffect(controller.focusRequests) {
        if (controller.focusRequests != controller.focusHonored) {
            controller.markFocusHonored()
            focusRequester.requestFocus()
        }
    }
    Surface(color = MaterialTheme.colorScheme.surfaceContainer, modifier = modifier.fillMaxWidth().testTag("richtext.findBar")) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 10.dp, vertical = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Icon(Icons.Filled.Search, contentDescription = null, tint = MaterialTheme.colorScheme.onSurfaceVariant)
            val textStyle = MaterialTheme.typography.bodyLarge.copy(color = LocalContentColor.current)
            BasicTextField(
                value = controller.query,
                onValueChange = { controller.query = it },
                singleLine = true,
                textStyle = textStyle,
                cursorBrush = SolidColor(MaterialTheme.colorScheme.primary),
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
                keyboardActions = KeyboardActions(onSearch = { controller.next() }),
                modifier = Modifier.weight(1f).focusRequester(focusRequester).testTag("richtext.findField"),
                decorationBox = { inner ->
                    if (controller.query.isEmpty()) {
                        Text(placeholder, style = textStyle, color = MaterialTheme.colorScheme.onSurfaceVariant)
                    }
                    inner()
                },
            )
            Text(
                controller.summary,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = 1,
                modifier = Modifier.testTag("richtext.findSummary"),
            )
            IconButton(onClick = { controller.previous() }, enabled = controller.ranges.isNotEmpty()) {
                Icon(Icons.Filled.KeyboardArrowUp, contentDescription = "Previous match")
            }
            IconButton(onClick = { controller.next() }, enabled = controller.ranges.isNotEmpty()) {
                Icon(Icons.Filled.KeyboardArrowDown, contentDescription = "Next match")
            }
            IconButton(onClick = { menuOpen = true }) {
                Icon(Icons.Filled.MoreVert, contentDescription = "Find options")
                DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                    OptionItem("Match case", controller.options.caseSensitive) {
                        controller.options = controller.options.copy(caseSensitive = it)
                    }
                    OptionItem("Whole words", controller.options.wholeWord) {
                        controller.options = controller.options.copy(wholeWord = it)
                    }
                    OptionItem("Match diacritics", controller.options.diacriticSensitive) {
                        controller.options = controller.options.copy(diacriticSensitive = it)
                    }
                    OptionItem("Regular expression", controller.options.regularExpression) {
                        controller.options = controller.options.copy(regularExpression = it)
                    }
                }
            }
            IconButton(onClick = { controller.dismiss() }) {
                Icon(Icons.Filled.Close, contentDescription = "Close find")
            }
        }
    }
}

@Composable
internal fun OptionItem(label: String, checked: Boolean, onChange: (Boolean) -> Unit) {
    DropdownMenuItem(
        text = { Text(label) },
        leadingIcon = { Checkbox(checked = checked, onCheckedChange = null) },
        onClick = { onChange(!checked) },
    )
}
